mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala

```diff
@@ -151,4 +167,64 @@ class StringIndexerModel private[ml] (
     val copied = new StringIndexerModel(uid, labels)
     copyValues(copied, extra)
   }
+
+  /**
+   * Return a model to perform the inverse transformation.
+   * Note: by default we keep the original columns during this transformation
+   * so the invert should only be needed if you do something beyond simply
+   * applying the original transformation.
+   */
+  def invert(): StringIndexerInvertModel = {
+    new StringIndexerInvertModel(uid, labels)
+      .setInputCol(getOutputCol)
+      .setOutputCol(getInputCol)
+  }
+}
+
+class StringIndexerInvertModel private[ml] (
+    override val uid: String,
+    labels: Array[String]) extends Model[StringIndexerInvertModel] with StringIndexerBase {
+
+  /** @group setParam */
+  def setInputCol(value: String): this.type = set(inputCol, value)
+
+  /** @group setParam */
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+
+  private val indexToLabel: OpenHashMap[Double, String] = {
```

Comment:
Do we expect use cases which are not 0-based indices?  I'd prefer this be stored as an Array of Strings (like labels already is, but pulled from metadata instead of given to the constructor).